lib/rules/no-else-return.js

```diff
@@ -33,7 +35,24 @@ module.exports = {
          * @returns {void}
          */
         function displayReport(node) {
-            context.report({ node, message: "Unnecessary 'else' after 'return'." });
+            context.report({
+                node,
+                message: "Unnecessary 'else' after 'return'.",
+                fix: fixer => {
+                    const sourceCode = context.getSourceCode();
+                    const startToken = sourceCode.getTokenByRangeStart(node.start);
+                    const elseToken = sourceCode.getTokenBefore(startToken);
+                    const source = sourceCode.getText(node);
+                    let fixedSource;
+
+                    if (startToken.type === "Punctuator") {
+                        fixedSource = source.substr(1, source.length - 2);
```

Comment:
Looks like you're right about these, my mistake. I missed the `Punctuator` check.

However, I noticed a few other errors (and double-checked them with the fixer this time):

* At the moment the `Punctuator` check is too loose -- it will treat any `Punctuator` node as if it was a block, even if it's not an opening curly. As a result, the following code causes a syntax error when fixed:

```js
function foo() {
  if (foo) return bar;
  else (foo).bar();
}


// gets fixed to:

function foo() {
  if (foo) return bar;
  foo).bar()
}
```

To fix this, I would recommend modifying the check to

```js
if (startToken.type === "Punctuator" && startToken.value === "{")
```

---

* Also, this will cause ASI issues if the statement in the `if` statement does not end in a semicolon. For example, the following code will be fixed incorrectly:

```js
function foo() {
  if (foo) return bar
  else { [1, 2, 3].map(foo) }
}

// gets fixed to

function foo() {
  if (foo) return bar
   [1, 2, 3].map(foo) // this is parsed as a computed property access rather than an array
}
```

To avoid this problem, I would recommend that the fixer avoid creating a fix when all of the following are true:

* The `if` statement does not have a block
* The body of the `if` statement does not end in a semicolon
* The first character of the `else` statement is an unsafe character for starting a line (`(`, `[`, `/`, `` ` ``, `+`, or `-`)

A similar check can be found in the [`no-lonely-if`](https://github.com/eslint/eslint/blob/ca1f8410fcb40a9bce286e22e23ee695159d15f7/lib/rules/no-lonely-if.js#L54-L69) fixer.

---

* In the third example from the previous comment, I think we should still use a workaround to avoid causing a syntax error, even if it would only occur when the user has very strange code.

```js
function foo() {
  if (foo) return bar
  else { baz } qux
}

// gets fixed to

function foo() {
  if (foo) return bar
   baz  qux
}
```

Similarly to the example above, this will also cause an ASI issue if the statement after the `else` starts with a character like `[`:

```js
function foo() {
  if (foo) return bar
  else { baz() }
  [1, 2, 3].map(foo)
}

// gets fixed to

function foo() {
  if (foo) return bar
   baz() 
  [1, 2, 3].map(foo) // this gets parsed as a computed property access rather than an array
}
```

To avoid this problem, I would recommend that the fixer avoid creating a fix when all of the following are true:

* The last statement in the `else` block does not end in a semicolon
* There is a statement after the `else` block, and it either (a) starts with an unsafe character (`(`, `[`, `/`, `` ` ``, `+`, or `-`), or (b) is on the same line as the last statement in the `else` block.

---

As an aside, hopefully it doesn't seem like I'm coming up with exceedingly unlikely code examples that no one would actually use. In general, we try to uphold the guarantee that an autofixer should never change the behavior of working code under any circumstances, so sometimes there are unfortunately a lot of edge cases for fixers like this (particularly when a fix involves removing a block statement).